File: src/_cpp/TowerDefenseBase/scenes/lab/Laboratory.cpp
//
//  IslandDefense
//
//  Created by Vladimir Tolmachev on 27.09.14.
//
//
#include "Laboratory.h"
#include "tower.h"
#include "UserData.h"
#include "ml/SmartScene.h"
#include "MapLayer.h"
#include "ScoreCounter.h"
#include "resources.h"
#include "ml/Audio/AudioEngine.h"
#include "configuration.h"
#include "ml/Language.h"
#include "ml/loadxml/xmlProperties.h"
#include "Tutorial.h"
#include "Achievements.h"

NS_CC_BEGIN

Laboratory::Laboratory()
: _scaleFactor( 1 )
{}

Laboratory::~Laboratory()
{}

bool Laboratory::init()
{
	do
	{
		auto dessize = Director::getInstance()->getOpenGLView()->getDesignResolutionSize();
		CC_BREAK_IF( !ScrollMenu::init() );
		CC_BREAK_IF( !NodeExt::init() );
		setCascadeOpacityEnabled( true );
		setCascadeColorEnabled( true );
		setKeyboardEnabled( true );

		NodeExt::load( "ini/laboratory", "lab.xml" );

		auto bg = getChildByName( "bg" );
		if( bg )
		{
			auto size = bg->getContentSize();
			size.width = std::min( dessize.width, size.width );
			size.height += bg->getPositionY();
			float sx = dessize.width / size.width;
			_scaleFactor = sx;
			bg->setScale( _scaleFactor );

			_zeroPosition.x = dessize.width / 2;
			_zeroPosition.y = size.height / 2 * _scaleFactor - 10;
			setPosition( _zeroPosition );
		}
        auto caption = getChildByName("text");
        if( caption )
        {
            caption->setPositionY(240.f);
            caption->setPositionX(0.f);
        }
        
        
		Size grid;
		Size content;
		std::list<std::string> towers;
		mlTowersInfo::shared().fetch( towers );
		for( auto & tower : towers )
		{
			auto item = buildItem( tower );
			addItem( item );

			item->setScale( _scaleFactor );

			grid = item->getContentSize();
			grid.width *= _scaleFactor;
			grid.height *= _scaleFactor;
			content.width += grid.width;
			content.height = grid.height;

			setIndicator( tower, false );
			setCost( tower );
			setParam( tower, false );
			setIcon( tower, false );
		}

		setAlignedStartPosition( Point( -_zeroPosition.x, -275 ) * _scaleFactor );
		setGrisSize( grid );
		align( 99 );

		Size scissor( dessize.width, 465 );
		setScissorRect( getAlignedStartPosition(), scissor * _scaleFactor );
		setScissorEnabled( true );
		setScrollEnabled( true );
		setContentSize( content );
		setAllowScrollByY( false );

		auto menu = getChildByName( "menu" );
		if( menu )
		{
			auto close = menu->getChildByName<MenuItem*>( "close" );
			close->setCallback( std::bind( &Laboratory::cb_close, this, std::placeholders::_1 ) );
			close->setScale( _scaleFactor );
			Point pos = close->getPosition();
			pos.y *= _scaleFactor;
			pos.x = dessize.width / 2 - 35;
			close->setPosition( pos );
		}

		fadeenter();
		return true;
	}
	while( false );
	return false;
}

void Laboratory::onKeyReleased( EventKeyboard::KeyCode keyCode, Event* event )
{
	if( keyCode == EventKeyboard::KeyCode::KEY_BACK )
		cb_close( nullptr );
}

MenuItemPointer Laboratory::buildItem( const std::string & tower )
{
	int level = UserData::shared().tower_upgradeLevel( tower );
	MenuItemPointer item = xmlLoader::load_node<MenuItem>( "ini/laboratory/item.xml" );
	item->setName( tower );
	item->setCallback( std::bind( &Laboratory::cb_select, this, std::placeholders::_1, tower ) );
	auto conteiner = item->getChildByName( "conteiner" );
	if( conteiner )
	{
		conteiner->setCascadeColorEnabled( true );
		conteiner->setCascadeOpacityEnabled( true );
		auto infonode = conteiner->getChildByName( "info" );
		auto mainnode = conteiner->getChildByName( "main" );
		mainnode->setCascadeColorEnabled( true );
		mainnode->setCascadeOpacityEnabled( true );

		auto caption = conteiner->getChildByName<Label*>( "name" );
		auto icon2 = conteiner->getChildByName<Sprite*>( "icon2" );
		auto icon_lock = conteiner->getChildByName<Sprite*>( "icon_locked" );

		auto indicator = mainnode->getChildByName( "indicator" );

		if( caption )
		{
			std::string name = tower + "_name";
			name = WORD( name );
			caption->setString( name );
		}
		if( icon2 )
		{
			xmlLoader::setProperty( icon2, xmlLoader::kImage, "images/laboratory/icon_" + tower + ".png" );
		}
		if( indicator )
		{
			indicator->setCascadeColorEnabled( true );
			indicator->setCascadeOpacityEnabled( true );
		}

		auto menuinfo = conteiner->getChildByName<Menu*>( "menu_info" );
		if( menuinfo )
		{
			auto info = menuinfo->getChildByName<MenuItem*>( "info" );
			info->setCallback( std::bind( &Laboratory::cb_info, this, std::placeholders::_1, tower ) );
		}

		auto menu = mainnode->getChildByName<Menu*>( "menu" );
		if( menu )
		{
			auto button = menu->getChildByName<MenuItem*>( "buy" );

			button->setName( tower );
			button->setCallback( std::bind( &Laboratory::cb_upgrade, this, std::placeholders::_1, tower ) );
		}
		auto menu_confirm = mainnode->getChildByName<Menu*>( "menu_confirm" );
		if( menu_confirm )
		{
			auto ok = menu_confirm->getChildByName<MenuItem*>( "confirm" );
			auto no = menu_confirm->getChildByName<MenuItem*>( "cancel" );

			ok->setName( tower );
			no->setName( tower );
			ok->setCallback( std::bind( &Laboratory::cb_confirm, this, std::placeholders::_1, tower ) );
			no->setCallback( std::bind( &Laboratory::cb_cancel, this, std::placeholders::_1, tower ) );
		}

		if( infonode )
		{
			auto text = infonode->getChildByName<Label*>( "text" );
			text->setString( mlTowersInfo::shared().get_desc( tower, 1 ) );
		}

		if( level == 0 )
		{
			item->setEnabled( false );
			item->setOpacity( 128 );
			menu_confirm->setEnabled( false );
			menu->setEnabled( false );
			menuinfo->setEnabled( false );
			menu->setOpacity( 128 );

			mainnode->setVisible( false );
			icon_lock->setVisible( true );
		}
	}

	return item;
}

void Laboratory::cb_select( Ref*, const std::string & tower )
{
	selectTower( tower );
}

void Laboratory::cb_info( Ref*sender, const std::string & tower )
{
	selectTower( tower );
	auto item = dynamic_cast<MenuItem*>(sender);
	assert( item );
	switchInfoBox( tower );
}

void Laboratory::cb_upgrade( Ref*, const std::string & tower )
{
	selectTower( tower );
	showConfirmMenu( tower, true );
	setParam( tower, true );
	setIcon( tower, true );
	setIndicator( tower, true );

	TutorialManager::shared( ).dispatch( "lab_clickupgrade" );
}

void Laboratory::cb_confirm( Ref*, const std::string & tower )
{
	int level = UserData::shared().tower_upgradeLevel( tower ) + 1;
	int cost = mlTowersInfo::shared().getCostLab( tower, level );
	int score = ScoreCounter::shared().getMoney( kScoreCrystals );

	if( score < cost )
	{
		if( k::configuration::useInapps )
		{
			if( k::configuration::useInapps && TutorialManager::shared().dispatch( "lab_haventgold" ) )
			{
				cb_close( nullptr );
				return;
			}
			else
			{
				SmartScene * scene = dynamic_cast<SmartScene*>(getScene( ));
				if( scene )
				{
					MapLayer::Pointer map( dynamic_cast<MapLayer*>(scene->getMainLayer( ).ptr( )) );
					if( map )
						map->cb_shop( nullptr );
				}
			}
		}
		else
		{
			cb_cancel(nullptr, tower);
			return;
		}
	}
	else
	{
		selectTower( "" );
		showConfirmMenu( tower, false );
		upgradeTower( tower );
		normalStateForAllItemExcept( "" );
		ScoreCounter::shared().subMoney( kScoreCrystals, cost, true );
		AudioEngine::shared().playEffect( kSoundLabUpgrade );
		UserData::shared().save();
	}
	setParam( tower, false );
	setIcon( tower, false );

	TutorialManager::shared( ).dispatch( "lab_clickconfirm" );
}

void Laboratory::cb_cancel( Ref*, const std::string & tower )
{
	selectTower( "" );
	showConfirmMenu( tower, false );
	normalStateForAllItemExcept( "" );
	setIndicator( tower, false );
}

void Laboratory::selectTower( const std::string & tower )
{
	normalStateForAllItemExcept( tower );
	unsigned count = getItemsCount();
	for( unsigned i = 0; i < count; ++i )
	{
		auto item = getItem( i );
		unsigned opacity = tower.empty() == false ?
			(item->getName() == tower ? 255 : 64) :
			255;
		item->setOpacity( opacity );
	}
}

void Laboratory::showConfirmMenu( const std::string & itemname, bool mode )
{
	int level = UserData::shared().tower_upgradeLevel( itemname );
	auto item = getItemByName( itemname );
	assert( item );
	auto conteiner = item->getChildByName( "conteiner" );
	auto mainnode = conteiner->getChildByName( "main" );
	auto menu0 = mainnode->getChildByName( "menu" );
	auto menu1 = mainnode->getChildByName( "menu_confirm" );

	menu0->setVisible( !mode && level != 5 );
	menu1->setVisible( mode );
}

void Laboratory::switchInfoBox( const std::string & itemname, bool forceHideInfo )
{
	int level = UserData::shared().tower_upgradeLevel( itemname );
	auto item = getItemByName( itemname );
	assert( item );
	auto conteiner = item->getChildByName( "conteiner" );
	auto infonode = conteiner->getChildByName( "info" );
	auto mainnode = conteiner->getChildByName( "main" );

	if( level > 0 )
	{
		bool infovisivle = forceHideInfo ? false : !infonode->isVisible();
		infonode->setVisible( infovisivle );
		mainnode->setVisible( !infovisivle );
	}
}

void Laboratory::upgradeTower( const std::string & tower )
{
	int level = UserData::shared().tower_upgradeLevel( tower ) + 1;
	level = std::min( level, 5 );
	UserData::shared().tower_upgradeLevel( tower, level );
	setIndicator( tower, false );
	setCost( tower );
	Achievements::shared().process( "lab_buyupgrade", 1 );
}

void Laboratory::setIndicator( const std::string & tower, bool nextLevel )
{
	int level = UserData::shared().tower_upgradeLevel( tower );
	if( nextLevel )
		level = std::min( level + 1, 5 );
	auto item = getItemByName( tower );
	auto conteiner = item->getChildByName( "conteiner" );
	auto main = conteiner ? conteiner->getChildByName( "main" ) : nullptr;
	auto indicator = main ? main->getChildByName( "indicator" ) : nullptr;
	if( indicator )
	{
		for( int i = 0; i < 5; ++i )
		{
			auto node = indicator->getChildByName( intToStr( i + 1 ) );
			if( node ) node->setVisible( i < level );
		}
		auto caption = indicator->getChildByName<Label*>( "caption" );
		if( level > 0 )
		{
			caption->setString( WORD("laboratory_tower_level") + intToStr( level ) );
		}
		else
		{
			caption->setString( "" );
		}
	}
	if( level == 5 && !nextLevel )
	{
		auto menu = main->getChildByName( "menu" );
		auto menu2 = main->getChildByName( "menu_confirm" );
		menu->setVisible( false );
		menu2->setVisible( false );
	}
}

void Laboratory::setCost( const std::string & tower )
{
	int level = UserData::shared().tower_upgradeLevel( tower );
	auto item = getItemByName( tower );
	int cost = mlTowersInfo::shared().getCostLab( tower, level + 1 );
	auto cost0 = dynamic_cast<Label*>(getNodeByPath( item, "conteiner/main/menu/" + tower + "/normal/cost" ));
	auto cost1 = dynamic_cast<Label*>(getNodeByPath( item, "conteiner/main/menu_confirm/cost" ));
	if( cost0 ) cost0->setString( intToStr( cost ) );
	if( cost1 ) cost1->setString( intToStr( cost ) );
}

void Laboratory::setParam( const std::string & tower, bool nextLevel )
{
	auto item = getItemByName( tower );
	auto conteiner = item->getChildByName( "conteiner" );
	if( !conteiner )return;
	auto mainnode = conteiner->getChildByName( "main" );
	if( !mainnode )return;

	int level = UserData::shared().tower_upgradeLevel( tower );
	if( nextLevel ) ++level;
	auto dmg = mainnode->getChildByName<Label*>( "dmg" );
	auto rng = mainnode->getChildByName<Label*>( "rng" );
	auto spd = mainnode->getChildByName<Label*>( "spd" );
	if( dmg )
	{
		std::string value = WORD("laboratory_tower_attack") + intToStr( mlTowersInfo::shared().get_dmg( tower, level ) );
		dmg->setString( value );
	}
	if( rng )
	{
		std::string value = WORD("laboratory_tower_range") + intToStr( mlTowersInfo::shared().get_rng( tower, level ) );
		rng->setString( value );
	}
	if( spd )
	{
		std::string value = WORD("laboratory_tower_speed") + intToStr( mlTowersInfo::shared().get_spd( tower, level ) );
		spd->setString( value );
	}
}

void Laboratory::setIcon( const std::string & tower, bool nextLevel )
{
	int level = UserData::shared().tower_upgradeLevel( tower );
	if( nextLevel ) ++level;
	level = std::min( std::max( 1, level ), 5 );

	std::string texture = "images/laboratory/icons/" + tower + intToStr( level ) + ".png";
	auto item = getItemByName( tower );
	auto conteiner = item->getChildByName( "conteiner" );
	if( !conteiner )return;
	auto icon = conteiner->getChildByName<Sprite*>( "icon" );
	if( icon ) xmlLoader::setProperty( icon, xmlLoader::kImage, texture );
}

void Laboratory::normalStateForAllItemExcept( const std::string & tower )
{
	for( unsigned i = 0; i < getItemsCount(); ++i )
	{
		auto item = getItem( i );
		if( item->getName() != tower )
		{
			showConfirmMenu( item->getName(), false );
			switchInfoBox( item->getName(), true );
			setParam( item->getName(), false );
			setIcon( item->getName(), false );
		}
	}
}

void Laboratory::cb_close( Ref* )
{
	auto func = [this]()
	{
		if( this->_callbackOnClosed )
			this->_callbackOnClosed();
		this->removeFromParent();
	};

	fadeexit();
	runAction( Sequence::createWithTwoActions(
		DelayTime::create( 0.5f ),
		CallFunc::create( func )
		) );
}

void Laboratory::fadeexit()
{
	static auto dessize = Director::getInstance()->getOpenGLView()->getDesignResolutionSize();
	auto action = EaseBackIn::create( MoveTo::create( 0.5f, _zeroPosition + Point( 0, -dessize.height ) ) );
	runAction( action );
	AudioEngine::shared().playEffect( kSoundShopHide );
}

void Laboratory::fadeenter()
{
	static auto dessize = Director::getInstance()->getOpenGLView()->getDesignResolutionSize();
	setPosition( _zeroPosition + Point( 0, -dessize.height ) );
	auto action = EaseBackOut::create( MoveTo::create( 0.5f, _zeroPosition ) );
	runAction( action );
	AudioEngine::shared().playEffect( kSoundShopShow );
}




NS_CC_END

File: src/_eu/scenes/lab/Laboratory.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/
/**TESTED**/
EU.Laboratory = EU.ScrollMenu.extend({
    scaleFactor: null,
    zeroPosition: null,
    self : null,
    ctor: function()
    {
        this._super();
        this.scaleFactor = 1;
        this.zeroPosition = cc.p(0,0);
        this.self = this;
        var dessize = cc.view.getDesignResolutionSize();
        this.setCascadeOpacityEnabled( true );
        this.setCascadeColorEnabled( true );
        //TODO: this.setKeyboardEnabled( true );

        this.load_str( "ini/laboratory/lab.xml" );

        var bg = this.getChildByName( "bg" );
        if( bg )
        {
            var size = bg.getContentSize();
            size.width = Math.min( dessize.width, size.width );
            size.height += bg.getPositionY();
            this.scaleFactor = dessize.width / size.width;
            bg.setScale( this.scaleFactor );

            this.zeroPosition.x = dessize.width / 2;
            this.zeroPosition.y = size.height / 2 * this.scaleFactor - 10;
            this.setPosition( this.zeroPosition );
        }
        var caption = this.getChildByName("text");
        if( caption )
        {
            caption.setPositionY(240);
            caption.setPositionX(0);
        }


        var grid = cc.size(0,0);
        var content = cc.size(0,0);
        var towers = [];
        EU.mlTowersInfo.fetch( towers );
        for( var i=0; i< towers.length; ++i )
        {
            var tower = towers[i];
            var item = this.buildItem( tower );
            this.addItem( item );

            item.setScale( this.scaleFactor );

            grid = item.getContentSize();
            grid.width *= this.scaleFactor;
            grid.height *= this.scaleFactor;
            content.width += grid.width;
            content.height = grid.height;

            this.setIndicator( tower, false );
            this.setCost( tower );
            this.setParam( tower, false );
            this.setIcon( tower, false );
        }

        this.setAlignedStartPosition( cc.p( -this.zeroPosition.x * this.scaleFactor, -275 * this.scaleFactor ) );
        this.setGrisSize( grid );
        this.align( 99 );

        var scissor = cc.size( dessize.width, 465 );
        var asp = this.getAlignedStartPosition();
        var visibleRect = cc.rect(asp.x, asp.y,scissor.width,scissor.height);
        this.setVisibleRect( visibleRect );
        this.setScrollEnabled( true );
        this.setTouchEnabled( true );
        this.setContentSize( content );
        this.setAllowScrollByY( false );

        var menu = this.getChildByName( "menu" );
        if( menu )
        {
            var close = menu.getChildByName( "close" );
            close.setCallback( this.cb_close, this );
            close.setScale( this.scaleFactor );
            var pos = close.getPosition();
            pos.y *= this.scaleFactor;
            pos.x = dessize.width / 2 - 35;
            close.setPosition( pos );
        }

        this.fadeenter();
        return true;
    },
    //TODO: keyboard
    //void onKeyReleased( EventKeyboard.KeyCode keyCode, Event* event )
    //{
    //    if( keyCode == EventKeyboard.KeyCode.KEY_BACK )
    //        cb_close( null );
    //}
    buildItem: function( tower )
    {
        var level = EU.UserData.tower_getUpgradeLevel( tower );
        var item = EU.xmlLoader.load_node_from_file( "ini/laboratory/item.xml" );
        item.setName( tower );
        item.setCallback( this.cb_select.bind(this, tower), this );
        var conteiner = item.getChildByName( "conteiner" );
        if( conteiner )
        {
            conteiner.setCascadeColorEnabled( true );
            conteiner.setCascadeOpacityEnabled( true );
            var infonode = conteiner.getChildByName( "info" );
            var mainnode = conteiner.getChildByName( "main" );
            mainnode.setCascadeColorEnabled( true );
            mainnode.setCascadeOpacityEnabled( true );
    
            var caption = conteiner.getChildByName( "name" );
            var icon2 = conteiner.getChildByName( "icon2" );
            var icon_lock = conteiner.getChildByName( "icon_locked" );
            var indicator = mainnode.getChildByName( "indicator" );
    
            if( caption )
            {
                var name = tower + "_name";
                name = EU.Language.string( name );
                caption.setString( name );
            }
            if( icon2 )
            {
                EU.xmlLoader.setProperty_int( icon2, EU.xmlKey.Image.int, "images/laboratory/icon_" + tower + ".png" );
            }
            if( indicator )
            {
                indicator.setCascadeColorEnabled( true );
                indicator.setCascadeOpacityEnabled( true );
            }
    
            var menuinfo = conteiner.getChildByName( "menu_info" );
            if( menuinfo )
            {
                var info = menuinfo.getChildByName( "info" );
                info.userdata = {};
                info.userdata.towername = tower;
                info.setCallback( this.cb_info.bind(this, tower), this );
            }
    
            var menu = mainnode.getChildByName( "menu" );
            if( menu )
            {
                var button = menu.getChildByName( "buy" );
    
                button.setName( tower );
                button.setCallback( this.cb_upgrade.bind(this,tower), this );
            }
            var menu_confirm = mainnode.getChildByName( "menu_confirm" );
            if( menu_confirm )
            {
                var ok = EU.Common.getNodeByPath( menu_confirm, "menu/confirm" );
                var no = EU.Common.getNodeByPath( menu_confirm, "menu/cancel" );
    
                ok.setName( tower );
                no.setName( tower );
                ok.setCallback( this.cb_confirm.bind(this, tower), this );
                no.setCallback( this.cb_cancel.bind(this, tower), this );
            }
    
            if( infonode )
            {
                var text = infonode.getChildByName( "text" );
                text.setString( EU.mlTowersInfo.get_desc( tower, 1 ) );
            }
    
            if( level == 0 )
            {
                item.setEnabled( false );
                item.setOpacity( 128 );
                menu_confirm.setEnabled( false );
                menu.setEnabled( false );
                menuinfo.setEnabled( false );
                menu.setOpacity( 128 );
    
                mainnode.setVisible( false );
                icon_lock.setVisible( true );
            }
        }
    
        return item;
    },
    cb_select: function( tower )
    {
        this.selectTower( tower );
    },
    cb_info: function( tower )
    {
        this.selectTower( tower );
        this.switchInfoBox( tower );
    },
    cb_upgrade: function( tower )
    {
        this.selectTower( tower );
        this.showConfirmMenu( tower, true );
        this.setParam( tower, true );
        this.setIcon( tower, true );
        this.setIndicator( tower, true );
    
        EU.TutorialManager.dispatch( "lab_clickupgrade" );
    },
    cb_confirm: function( tower )
    {
        var level = EU.UserData.tower_getUpgradeLevel( tower ) + 1;
        var cost = EU.mlTowersInfo.getCostLab( tower, level );
        var score = EU.ScoreCounter.getMoney( EU.kScoreCrystals );
    
        if( score < cost )
        {
            if( EU.k.useInapps )
            {
                if( EU.k.useInapps && EU.TutorialManager.dispatch( "lab_haventgold" ) )
                {
                    this.cb_close( null );
                    return;
                }
                else
                {
                    var scene = EU.Common.getSceneOfNode(this);
                    if( scene )
                    {
                        var layer = scene.getMainLayer();
                        if( layer instanceof EU.MapLayer)
                            map.cb_shop();
                    }
                }
            }
            else
            {
                this.cb_cancel(null, tower);
                return;
            }
        }
        else
        {
            this.selectTower( "" );
            this.showConfirmMenu( tower, false );
            this.upgradeTower( tower );
            this.normalStateForAllItemExcept( "" );
            EU.ScoreCounter.subMoney( EU.kScoreCrystals, cost, true );
            //TODO: audio
            //AudioEngine.playEffect( kSoundLabUpgrade );
            EU.UserData.save();
        }
        this.setParam( tower, false );
        this.setIcon( tower, false );
    
        EU.TutorialManager.dispatch( "lab_clickconfirm" );
    },
    cb_cancel: function( tower )
    {
        this.selectTower( "" );
        this.showConfirmMenu( tower, false );
        this.normalStateForAllItemExcept( "" );
        this.setIndicator( tower, false );
    },    
    selectTower: function( tower )
    {
        this.normalStateForAllItemExcept( tower );
        var count = this.getItemsCount();
        for( var i = 0; i < count; ++i )
        {
            var item = this.getItem( i );
            var opacity = tower.length > 0 ?
                (item.getName() == tower ? 255 : 255) :
                255;
            item.setOpacity( opacity );
        }
    },    
    showConfirmMenu: function( itemname, mode )
    {
        var level = EU.UserData.tower_getUpgradeLevel( itemname );
        var item = this.getItemByName( itemname );
        EU.assert( item );
        var conteiner = item.getChildByName( "conteiner" );
        var mainnode = conteiner.getChildByName( "main" );
        var menu0 = mainnode.getChildByName( "menu" );
        var menu1 = mainnode.getChildByName( "menu_confirm" );
    
        menu0.setVisible( !mode && level != 5 );
        menu1.setVisible( mode );
    },
    switchInfoBox: function( itemname, forceHideInfo )
    {
        var level = EU.UserData.tower_getUpgradeLevel( itemname );
        var item = this.getItemByName( itemname );
        EU.assert( item );
        var conteiner = item.getChildByName( "conteiner" );
        var infonode = conteiner.getChildByName( "info" );
        var mainnode = conteiner.getChildByName( "main" );
    
        if( level > 0 )
        {
            var infovisivle = forceHideInfo ? false : !infonode.isVisible();
            infonode.setVisible( infovisivle );
            mainnode.setVisible( !infovisivle );
        }
    },    
    upgradeTower: function( tower )
    {
        var level = EU.UserData.tower_getUpgradeLevel( tower ) + 1;
        level = Math.min( level, 5 );
        EU.UserData.tower_setUpgradeLevel( tower, level );
        this.setIndicator( tower, false );
        this.setCost( tower );
        //TODO: Achievments
        //Achievements.process( "lab_buyupgrade", 1 );
    },    
    setIndicator: function( tower, nextLevel )
    {
        var level = EU.UserData.tower_getUpgradeLevel( tower );
        if( nextLevel )
            level = Math.min( level + 1, 5 );
        var item = this.getItemByName( tower );
        var conteiner = item.getChildByName( "conteiner" );
        var main = conteiner ? conteiner.getChildByName( "main" ) : null;
        var indicator = main ? main.getChildByName( "indicator" ) : null;
        if( indicator )
        {
            for( var i = 0; i < 5; ++i )
            {
                var node = indicator.getChildByName( (i + 1).toString() );
                if( node ) node.setVisible( i < level );
            }
            var caption = indicator.getChildByName( "caption" );
            if( level > 0 )
            {
                caption.setString( EU.Language.string("laboratory_tower_level") + level.toString() );
            }
            else
            {
                caption.setString( "" );
            }
        }
        if( level == 5 && !nextLevel )
        {
            var menu = main.getChildByName( "menu" );
            var menu2 = main.getChildByName( "menu_confirm" );
            menu.setVisible( false );
            menu2.setVisible( false );
        }
    },    
    setCost: function( tower )
    {
        var level = EU.UserData.tower_getUpgradeLevel( tower );
        var item = this.getItemByName( tower );
        var cost = EU.mlTowersInfo.getCostLab( tower, level + 1 );
        var cost0 = EU.Common.getNodeByPath( item, "conteiner/main/menu/" + tower + "/normal/cost" );
        var cost1 = EU.Common.getNodeByPath( item, "conteiner/main/menu_confirm/cost" );
        if( cost0 ) cost0.setString( cost );
        if( cost1 ) cost1.setString( cost );
    },    
    setParam: function( tower, nextLevel )
    {
        var item = this.getItemByName( tower );
        var conteiner = item.getChildByName( "conteiner" );
        if( !conteiner )return;
        var mainnode = conteiner.getChildByName( "main" );
        if( !mainnode )return;
    
        var level = EU.UserData.tower_getUpgradeLevel( tower );
        if( nextLevel ) ++level;
        var dmg = mainnode.getChildByName( "dmg" );
        var rng = mainnode.getChildByName( "rng" );
        var spd = mainnode.getChildByName( "spd" );
        var value;
        if( dmg )
        {
            value = EU.Language.string("laboratory_tower_attack") + Math.round(EU.mlTowersInfo.get_dmg( tower, level ));
            dmg.setString( value );
        }
        if( rng )
        {
            value = EU.Language.string("laboratory_tower_range") + Math.round(EU.mlTowersInfo.get_rng( tower, level ));
            rng.setString( value );
        }
        if( spd )
        {
            value = EU.Language.string("laboratory_tower_speed") + Math.round(EU.mlTowersInfo.get_spd( tower, level ));
            spd.setString( value );
        }
    },    
    setIcon: function( tower, nextLevel )
    {
        var level = EU.UserData.tower_getUpgradeLevel( tower );
        if( nextLevel ) ++level;
        level = Math.min( Math.max( 1, level ), 5 );
    
        var texture = "images/laboratory/icons/" + tower + level + ".png";
        var item = this.getItemByName( tower );
        var conteiner = item.getChildByName( "conteiner" );
        if( !conteiner )return;
        var icon = conteiner.getChildByName( "icon" );
        if( icon ) EU.xmlLoader.setProperty_int( icon, EU.xmlKey.Image.int, texture );
    },
    normalStateForAllItemExcept: function( tower )
    {
        for( var i = 0; i < this.getItemsCount(); ++i )
        {
            var item = this.getItem( i );
            if( item.getName() != tower )
            {
                this.showConfirmMenu( item.getName(), false );
                this.switchInfoBox( item.getName(), true );
                this.setParam( item.getName(), false );
                this.setIcon( item.getName(), false );
            }
        }
    },
    cb_close: function()
    {
        this.fadeexit();
        this.runAction( cc.sequence(
            cc.delayTime( 0.5 ),
            cc.callFunc( this.removeFromParent, this )) );
    },
    fadeexit: function()
    {
        var dessize = cc.view.getDesignResolutionSize();
        var action = cc.moveTo( 0.5, EU.Common.pointAdd(this.zeroPosition, cc.p( 0, -dessize.height ) ) ).easing(cc.easeBackIn());
        this.runAction( action );
        //TODO: audio
        //AudioEngine.playEffect( kSoundShopHide );
    },
    fadeenter: function()
    {
        var dessize = cc.view.getDesignResolutionSize();
        this.setPosition( EU.Common.pointAdd(this.zeroPosition, cc.p( 0, -dessize.height ) ) );
        var action = cc.moveTo( 0.5, this.zeroPosition ).easing(cc.easeBackOut());
        this.runAction( action );
        //TODO: audio
        //AudioEngine.playEffect( kSoundShopShow );
    }

});
EU.NodeExt.call(EU.Laboratory.prototype);
